megatron/rl/rl_utils.py

```diff
@@ -1326,751 +838,235 @@ def prepare_data_for_update(
-                print_rank_0(
-                    f"  - Microbatches per step: {new_num_microbatches} (was {old_num_microbatches})"
-                )
-
-                for step in range(min(3, total_steps)):
-                    start_idx = step * bins_per_rank_per_step
-                    end_idx = min(start_idx + bins_per_rank_per_step, num_bins_this_rank)
-                    step_bins = end_idx - start_idx
-
-                    actual_seqs = sum(
-                        len(my_bin_seq_indices[bin_idx])
-                        for bin_idx in range(start_idx, end_idx)
-                        if bin_idx < len(my_bin_seq_indices)
-                    )
-                    est_global_seqs = actual_seqs * dp_world_size
-                    print_rank_0(
-                        f"  - Step {step + 1}: {step_bins} bins, ~{est_global_seqs} sequences globally"
+        if args.rl_use_sequence_packing:
+            with nvtx_range("pack_logprobs", time=True):
+                # Store logprobs on gpu in packing context
+                # Since PackingContext is a dataclass, we add these as new attributes
+                packing_context.old_logprobs = old_logprobs.cuda()
+                packing_context.ref_logprobs = ref_logprobs.cuda()
+
+                if inference_logprobs is not None and args.rl_inference_logprobs_is_correction:
+                    # Pack the inference logprobs using the helper function
+                    packed_inference_logprobs = pack_inference_logprobs(
+                        inference_logprobs=packing_context.original_inference_logprobs,
+                        packing_info=packing_context.packing_info,
+                        generation_masks=packing_context.original_generation_masks,
+                        bin_size=args.seq_length,
                     )
 
-                if total_steps > 3:
-                    print_rank_0(f"  - ... ({total_steps - 3} more steps)")
+                    # Store packed inference logprobs in packing context
+                    packing_context.packed_inference_logprobs = packed_inference_logprobs.cuda()
+                    packing_context.has_inference_logprobs = True
+        else:
+            with nvtx_range("align_inference_logprobs", time=True):
+                if inference_logprobs is not None:
+                    inference_logprobs = align_unpacked_inference_logprobs(
+                        inference_logprobs=inference_logprobs,
+                        old_logprobs_for_data=old_logprobs,
+                        generation_masks=generation_masks,
+                        group_stats=group_stats,
+                    )
+                    # We run the above to fill in the inference/train side mismatch stats.
+                    # We do the above for logging purposes.
+                    # Nullify logprobs if not used in IS correction,
+                    if not args.rl_inference_logprobs_is_correction:
```

Comment:
I think this brings back the bug we had before. If we do not use is correction, we still want to keep the inference logprobs for logging purposes. I fixed it for the no-sequence-packing case, but this will break it again.

I think, we need to fix it. For both, packing or not packing, we need inference logprobs to log the mismatch between train and inference side.